### ISSUE TITLE
Failed to use QueryRewriter as a separate bean

### ISSUE THREAD

**Issue by sergey-morenets** [spring-projects/spring-data-jpa#2539]
Hi

I tried to use new QueryRewriter functionality as described in https://spring.io/blog/2022/05/02/ever-wanted-to-rewrite-a-query-in-spring-data-jpa

And it was mentioned that "While you can write your QueryRewriter as a separate bean, it’s also possible to put right inside the repository where it’s used!"

I created and registered my own Spring bean and unfortunately it's not picked up by Spring Data:

```
public class DefaultQueryRewriter implements QueryRewriter {

	@Override
	public String rewrite(String query, Sort sort) {
		
		return query.replaceAll("FROM Product", "FROM Products");
	}
}

	@Bean
	public QueryRewriter queryRewriter() {
		return new DefaultQueryRewriter();
	}
```

So the only option is to specify new class in the @Query annotation:

```
	@Query(value="from Product where id=:id",
			queryRewriter = DefaultQueryRewriter.class)
	Product findWithId(@Param("id") int id);
```

Spring Data JPA 3.0.0 M4

**Comment by gregturn** [comment]
@sergey-morenets what happens if you tweak things like this:

```java
public class DefaultQueryRewrite implements QueryRewriter { ...your rewrite() method... }

@Configuration
public class CustomConfig {
	@Bean
	public QueryRewriter queryRewriter() {
		return new DefaultQueryRewriter();
	}
}
```

**Comment by sergey-morenets** [comment]
Hi @gregturn 

Yes, that's exactly what I did (this is simplified Java-based configuration):

```
@Configuration
@EnableJpaAuditing
public class PersistenceConfig {

	@Bean
	public QueryRewriter queryRewriter() {
		return new DefaultQueryRewriter();
	}
}
```

**Comment by rishiraj88** [comment]
Great!

**Comment by gregturn** [comment]
Okay, I'm perplexed. Can you boil this down to a Spring Boot project and link to it so we can scrutinize it closer?

**Comment by sergey-morenets** [comment]
Hi @gregturn 

Yes, please see the link to my repository: https://github.com/sergey-morenets/query-rewriter

This is actually a project generated by Spring Starter where I added simplified model, configuration, repository:

```
public interface ProductRepository extends JpaRepository<Product, Integer> {

    @Query(value = "SELECT * FROM PRODUCTS", nativeQuery = true)
    List<Product> findProducts();
}

query rewriter:

```
public class DefaultQueryRewriter implements QueryRewriter {

    @Override
    public String rewrite(String query, Sort sort) {

        return query.replaceAll("FROM PRODUCTS", "FROM Product");
    }
}
```

```

and test:

```
@DataJpaTest
class DemoApplicationTests {

	@Autowired
	ProductRepository productRepository;

	@Test
	void findProducts_noProductDefined_returnsEmtpty(ApplicationContext context) {
		DefaultQueryRewriter bean = context.getBean(DefaultQueryRewriter.class);
		assertNotNull(bean);

		List<Product> products = productRepository.findProducts();
		assertTrue(products.isEmpty());
	}
}
```

And if I run the test it fails with error message: 

```
Caused by: org.h2.jdbc.JdbcSQLSyntaxErrorException: Table "PRODUCTS" not found; SQL statement:
SELECT * FROM PRODUCTS [42102-212]
```

However if I specify the query rewriter in the ProductRepository method then test passes:

```
    @Query(value = "SELECT * FROM PRODUCTS", nativeQuery = true, queryRewriter = DefaultQueryRewriter.class)
    List<Product> findProducts();
```

**Comment by gregturn** [comment]
1. I've created my own instance of a simple variant that seems to work. Check it out at https://github.com/gregturn/spring-data-query-rewriter. It basically has a barebones, 1-table setup, and only logs the query that was found. Nothing more complicated than that.
3. @sergey-morenets reading through, I'm not sure why you change the name of the table from PRODUCT to PRODUCTS. Also, I wouldn't mix `nativeQuery=true` into this, until we've nailed down where things are NOT working.
4. I'm also skeptical of stirring `@DataJpaTest` in either, since that COULD introduce a wrinkle we're also not aware of.
5. Do you have overrides on table creation? Spring Boot should, by default, have a `create-drop` policy for embeddable databases (H2 and HSQL). For other database types it will switch to `none`.

To unravel all this, we may have to tweak one setting at a time, and I fear we have five things varying.

**Comment by gregturn** [comment]
@sergey-morenets I poked at your example, and it appears the entity definition was the culprit. With that fixed, I was certainly able to get your query rewriter to work.

https://github.com/sergey-morenets/query-rewriter/pull/1

**Comment by sergey-morenets** [comment]
Hi @gregturn 

Thank you for the example. I've read in the documentation:

`This example shows both a native (pure SQL) rewriter as well as a JPQL query, both leveraging the same QueryRewriter. In this scenario, Spring Data JPA will look for a bean registered in the application context of the corresponding type.`

So my question is that possible to define Spring bean (QueryRewriter) that will be automatically applied to all queries even if it's not specified in @Query annotation ?

**Comment by gregturn** [comment]
No. You MUST specify it in the `@Query` annotation. The shortcut of having the repository ALSO extend this interface is simply to cut things down to one place to hold both your queries AND your rewriters. It's not meant to automatically it everywhere.

Trying to apply it wider than this seems like a corner case I'm not sure we want to support (yet).

**Comment by gregturn** [comment]
And yes, I forgot that I wrote unit tests to verify this works with both native SQL and JPQL.

I also have deduced that your reproducer tries to use the rewriter to "fix" the query. I think it fails because when the app starts up, Spring Data repositories get routed through a validation phase. However, the query rewriter is lazy and doesn't do its thing until invoked.

So I believe your whole example failed due to having an invalid query inside `@Query`, which worked when I altered the table name, Fixing the query to be correct AT STARTUP would also have mended the situation.
